File: Dynamic-Programming/583. Delete Operation for Two Strings/deleteOperationforTwoStrings.java
import java.util.HashMap;
// Time Complexity: O(m * n)
class Solution {
    public int minDistance(String word1, String word2) {
        this.word1 = word1;
        this.word2 = word2;
        return min(word1.length(), word2.length());
    }
    
    String word1;
    String word2;
    
    HashMap<String, Integer> memo = new HashMap<>();
    
    int min(int size1, int size2) {
        // base case
        // from right to left, so base case can be 0
        if (size1 == 0 && size2 == 0) {
            return 0;
        } else if (size1 == 0) {
            return size2;
        } else if (size2 == 0) {
            return size1;
        }
        
        //size1 > 0 and size2 > 0;
        String key = size1 + "." + size2;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        
        int solution = 0;
        char c1 = word1.charAt(size1 - 1);
        char c2 = word2.charAt(size2 - 1);
        if (c1 == c2) {
            solution = min(size1 - 1, size2 - 1);
            // the values are brought up
            System.out.println("solution");
            System.out.println(solution);
        } else {
            int option1 = min(size1 - 1, size2) + 1;
            int option2 = min(size1, size2 - 1)  + 1;
            solution = Math.min(option1, option2);
            // print out size1 and size2
            // the values are brought up
            System.out.println("o1");
            System.out.println(option1);
            System.out.println("o2");
            System.out.println(option2);
            System.out.println("sol");
            System.out.println(solution);
        }
        
        memo.put(key, solution);
        return solution;
    }
}

File: Dynamic-Programming/583. Delete Operation for Two Strings/deleteOperationforTwoStrings.py
#  Time Complexity: O(m * n)
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = [[0 for j in range(len(word2) + 1)] for i in range(len(word1) + 1)]
        
        for i in range(len(word1) - 1, -1, -1):
            for j in range(len(word2) - 1, -1, -1):
                if word1[i] == word2[j]:
                    dp[i][j] = 1 + dp[i + 1][j + 1]
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j + 1])
        
        return len(word1) + len(word2) - dp[0][0] * 2
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m = len(word1)
        n = len(word2)
        dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        
        for i in range(1, m + 1):
            dp[0][i] += dp[0][i - 1] + 1
        
        for i in range(1, n + 1):
            dp[i][0] += dp[i - 1][0] + 1
        
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if word1[j - 1] == word2[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i][j - 1], dp[i - 1][j]) + 1

        return dp[-1][-1]
        
        
        #    "" s e a
        # "" 0  1 2 3
        # e  1  2 1 2
        # a  2  3 2 1  
        # t  3  4 3 2